src/Stage/InferTypes/GenerateEquations.elm

```diff
@@ -412,3 +456,165 @@ generateArgumentUsageEquations argumentId usages =
+            in
+            ( equals type_ (Type.Tuple fstType sndType)
+                :: fstEquations
+                ++ sndEquations
+            , id2
+            )
+
+        Typed.PTuple3 fst snd trd ->
+            let
+                ( _, fstType ) =
+                    Located.unwrap fst
+
+                ( _, sndType ) =
+                    Located.unwrap snd
+
+                ( _, trdType ) =
+                    Located.unwrap trd
+
+                ( fstEquations, id1 ) =
+                    generatePatternEquations currentId fst
+
+                ( sndEquations, id2 ) =
+                    generatePatternEquations id1 snd
+
+                ( trdEquations, id3 ) =
+                    generatePatternEquations id2 trd
+            in
+            ( equals type_ (Type.Tuple3 fstType sndType trdType)
+                :: fstEquations
+                ++ sndEquations
+                ++ trdEquations
+            , id3
+            )
+
+        Typed.PList_ items ->
+            {- The list type parameter needs extra ID so that we can
+               bind the items' types to it... so we create one here.
+            -}
+            let
+                id1 =
+                    currentId + 1
+
+                listParamType =
+                    Type.Var currentId
+
+                ( bodyEquations, id2 ) =
+                    List.foldr
+                        (\item ( acc, currentId_ ) ->
+                            let
+                                ( _, itemType ) =
+                                    Located.unwrap item
+
+                                ( equations, nextId ) =
+                                    generatePatternEquations currentId_ item
+                            in
+                            ( equals itemType listParamType
+                                :: equations
+                                ++ acc
+                            , nextId
+                            )
+                        )
+                        ( [], id1 )
+                        items
+            in
+            ( -- for expression `[ a, b, c ]`
+              -- the `x` in `List x` type and types of all the items are the same
+              equals type_ (Type.List listParamType)
+                :: bodyEquations
+            , id2
+            )
+
+        Typed.PCons left right ->
+            let
+                ( _, leftType ) =
+                    Located.unwrap left
+
+                ( _, rightType ) =
+                    Located.unwrap right
+
+                ( leftEquations, id1 ) =
+                    generatePatternEquations currentId left
+
+                ( rightEquations, id2 ) =
+                    generatePatternEquations id1 right
+            in
+            ( -- For expression a :: [ b ]:
+              [ equals rightType (Type.List leftType) -- type of b is a List a
+              , equals type_ rightType -- a :: [ b ] is a List b
```

Comment:
I get what you mena by these comments but they are confusing to me. I guess I'd skip the `[ b ]` and write `b` instead.

```suggestion
            ( -- For expression a :: b:
              [ equals rightType (Type.List leftType) -- type of `b` is `List a`
              , equals type_ rightType -- type of `a :: b` is `List a`
```
Still kinda weird (are `a` and `b` values or types?) but seems a little bit better to me.